File: RubyInRails/app/controllers/users_controller.rb
class UsersController < ApplicationController
  def create
    user = UserService.create(user_params)
    if user
      render json: user, status: :created
    else
      render json: { error: "Invalid input" }, status: :bad_request
    end
  rescue ActiveRecord::RecordNotUnique
    render json: { error: "User already exists" }, status: :bad_request
  end

  def index
    users = UserService.list_all
    render json: users
  end

  def show
    user = UserService.get(params[:id])
    render json: user
  end

  def update
    user = UserService.update(params[:id], user_params)
    render json: user
  end

  def destroy
    UserService.delete(params[:id])
    head :no_content
  end

  private

  def user_params
    params.require(:user).permit(:username, :email)
  rescue ActionController::ParameterMissing
    # Handle cases where json root is missing or raw json is sent without root
    params.permit(:username, :email)
  end
end

File: JuliaGenie/src/controllers/UsersController.jl
module UsersController

using Genie.Requests
using Genie.Renderer.Json
using ..UsersService

function create()
  try
      payload = jsonpayload()
      username = get(payload, "username", "")
      email = get(payload, "email", "")

      user = UsersService.create_user(username, email)
      json(user, status = 201)
  catch e
      json(Dict("error" => string(e)), status = 400)
  end
end

function index()
  users = UsersService.list_users()
  json(users)
end

function show()
  id = parse(Int, params(:id))
  user = UsersService.get_user(id)
  if user === nothing
      json(Dict("error" => "User not found"), status = 404)
  else
      json(user)
  end
end

function update()
  try
      id = parse(Int, params(:id))
      payload = jsonpayload()
      username = get(payload, "username", "")

      user = UsersService.update_user(id, username)
      if user === nothing
          json(Dict("error" => "User not found"), status = 404)
      else
          json(user)
      end
  catch e
      json(Dict("error" => string(e)), status = 400)
  end
end

function delete()
  id = parse(Int, params(:id))
  UsersService.delete_user(id)
  Genie.Renderer.Json.json(nothing, status = 204)
end

end
